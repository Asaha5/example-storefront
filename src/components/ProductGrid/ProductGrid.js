import React, { Component } from "react";
import PropTypes from "prop-types";
import Grid from "@material-ui/core/Grid";
import { ResultCard } from "@appbaseio/reactivesearch";
import { withStyles } from "@material-ui/core/styles";
import track from "lib/tracking/track";
import trackProductClicked from "lib/tracking/trackProductClicked";
import PageSizeSelector from "components/PageSizeSelector";
import SortBySelector from "components/SortBySelector";

const styles = (theme) => ({
  filters: {
    justifyContent: "flex-end",
    marginBottom: theme.spacing.unit * 2
  },
  resultCard: {
    "& .resultImage": {
      boxShadow: "none"
    }
  }
});

export const searchResultCardProps = {
  componentId: "searchResultCard",
  dataField: "product.title",
  size: 20,
  onData: (res) =>
    ({
      description: res.product.vendor,
      image: `http://localhost:3000/${res.product.media[0].URLs.small}`,
      title: (
        <div>
          {res.product.title}
          {res.product.pricing.USD.displayPrice}
        </div>
      ),
      url: `/product/${res.product.slug}`
    }),
  pagination: true,
  react: {
    and: ["SearchInput"]
  },
  target: "_self"
};

@withStyles(styles, { name: "SkProductGrid" })
@track()
export default class ProductGrid extends Component {
  static propTypes = {
    catalogItems: PropTypes.arrayOf(PropTypes.object),
    classes: PropTypes.object,
    // currencyCode: PropTypes.string.isRequired,
    initialSize: PropTypes.object,
    isLoadingCatalogItems: PropTypes.bool,
    pageInfo: PropTypes.shape({
      startCursor: PropTypes.string,
      endCursor: PropTypes.string,
      hasNextPage: PropTypes.bool,
      hasPreviousPage: PropTypes.bool,
      loadNextPage: PropTypes.func,
      loadPreviousPage: PropTypes.func
    })
    // pageSize: PropTypes.number.isRequired,
    // setPageSize: PropTypes.func.isRequired,
    // setSortBy: PropTypes.func.isRequired,
    // sortBy: PropTypes.string.isRequired
  };

  renderFilters() {
    const { classes, pageSize, setPageSize, setSortBy, sortBy } = this.props;

    return (
      <Grid container spacing={8} className={classes.filters}>
        <Grid item>
          <PageSizeSelector pageSize={pageSize} onChange={setPageSize} />
        </Grid>
        <Grid item>
          <SortBySelector sortBy={sortBy} onChange={setSortBy} />
        </Grid>
      </Grid>
    );
  }

  @trackProductClicked()
  onItemClick = (event, product) => {} // eslint-disable-line no-unused-vars

  render() {
    const { classes } = this.props;

    return (
      <ResultCard
        {
        ...searchResultCardProps
        }
        className={classes.resultCard}
        innerClass={{ listItem: "resultImage" }}
      />
    );
  }
}
