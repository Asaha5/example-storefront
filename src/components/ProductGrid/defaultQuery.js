export default `
{
  productSearchPagination(
    query: { match_all:{} }, 
    page: PAGE,
    perPage: PER_PAGE
  ) 
  {
    count
    items {
      _score
      _source {
        id
        product {
          description
          media
          pricing
          slug
          title
          vendor
          variants {
            _id
            title
            options {
              _id
              optionTitle
            }
          }
        }
      }
    }
  }
}
`;
