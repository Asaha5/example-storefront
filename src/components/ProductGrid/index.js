export { default } from "./ProductGrid";
export { searchResultCardProps } from "./ProductGrid";
