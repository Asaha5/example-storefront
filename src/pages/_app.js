import NextApp, { Container } from "next/app";
import React from "react";
import { ReactiveBase } from "@appbaseio/reactivesearch";
import { ThemeProvider as RuiThemeProvider } from "styled-components";
import { StripeProvider } from "react-stripe-elements";
import { MuiThemeProvider } from "@material-ui/core/styles";
import CssBaseline from "@material-ui/core/CssBaseline";
import JssProvider from "react-jss/lib/JssProvider";
import { ComponentsProvider } from "@reactioncommerce/components-context";
import get from "lodash.get";
import getConfig from "next/config";
import track from "lib/tracking/track";
import dispatch from "lib/tracking/dispatch";
import withApolloClient from "lib/apollo/withApolloClient";
import withShop from "containers/shop/withShop";
import withViewer from "containers/account/withViewer";
import withTags from "containers/tags/withTags";
import Layout from "components/Layout";
import withMobX from "lib/stores/withMobX";
import rootMobXStores from "lib/stores";
import components from "../lib/theme/components";
import getPageContext from "../lib/theme/getPageContext";
import componentTheme from "../lib/theme/componentTheme";

const { publicRuntimeConfig } = getConfig();
const matchAllDefaultQuery = `
{
  productSearchConnection(
    query: { match_all:{} }, 
    sort: _score, 
    first: 1
  ) 
  {
    _shards {
      successful,
      failed
      total
    }
    count
    max_score
    took
    timed_out
    pageInfo {
      hasNextPage
      hasPreviousPage
      startCursor
      endCursor
    }
    edges {
      cursor
      node {
        _score
        _source {
          product {
            description
            media
            pricing
            slug
            title
            vendor
          }
        }
      }
    }
  }
}
`;

@withApolloClient
@withMobX
@withShop
@withViewer
@withTags
@track({}, { dispatch })
export default class App extends NextApp {
  static async getInitialProps({ Component, ctx }) {
    let pageProps = {};

    if (Component.getInitialProps) {
      pageProps = await Component.getInitialProps(ctx);
    }

    return { pageProps };
  }

  constructor(props) {
    super(props);
    this.pageContext = getPageContext();
    this.state = { stripe: null };
  }

  pageContext = null;

  componentDidMount() {
    // Fetch and update auth token in auth store
    rootMobXStores.cartStore.setAnonymousCartCredentialsFromLocalStorage();
    rootMobXStores.keycloakAuthStore.setTokenFromLocalStorage();

    // Remove the server-side injected CSS.
    const jssStyles = document.querySelector("#jss-server-side");
    if (jssStyles && jssStyles.parentNode) {
      jssStyles.parentNode.removeChild(jssStyles);
    }

    const { stripePublicApiKey } = publicRuntimeConfig;
    if (stripePublicApiKey && window.Stripe) {
      // eslint-disable-next-line react/no-did-mount-set-state
      this.setState({ stripe: window.Stripe(stripePublicApiKey) });
    }
  }

  transformSearchRequest = (request) => {
    const [preference, elasticQuery] = request.body.split("\n");
    // console.log("elasticQuery", JSON.parse(elasticQuery));
    console.log("preference", preference);


    // Get query from sensor component, i.e. SearchInput component
    let graphqlQuery = get(JSON.parse(elasticQuery), "query.bool.must[0].bool.must.graphqlQuery");
    if (!graphqlQuery) {
      // Get query from a results component, i.e. ResultCard
      graphqlQuery = get(JSON.parse(elasticQuery), "query.bool.must[0].bool.must[0].graphqlQuery");
      // Get initial query, "match_all: {}"
      if (!graphqlQuery) {
        // TODO: add dynamic values for first, and cursor args
        graphqlQuery = matchAllDefaultQuery;
      }
    }

    request.body = JSON.stringify(graphqlQuery);

    return request;
  }

  render() {
    const { Component, pageProps, shop, viewer, ...rest } = this.props;
    const { route } = this.props.router;
    const { stripe } = this.state;

    return <Container>
      <ComponentsProvider value={components}>
        <JssProvider registry={this.pageContext.sheetsRegistry} generateClassName={this.pageContext.generateClassName}>
          <ReactiveBase
            app="reaction.cdc.reaction.catalog.json-gen1"
            transformRequest={this.transformSearchRequest}
            url="http://graphql-proxy.search-api.reaction.localhost:9201"
          >
            <RuiThemeProvider theme={componentTheme}>
              <MuiThemeProvider theme={this.pageContext.theme} sheetsManager={this.pageContext.sheetsManager}>
                <CssBaseline />
                {route === "/checkout" || route === "/login" ? <StripeProvider stripe={stripe}>
                  <Component pageContext={this.pageContext} shop={shop} {...rest} {...pageProps} />
                </StripeProvider> : <Layout shop={shop} viewer={viewer}>
                  <Component pageContext={this.pageContext} shop={shop} {...rest} {...pageProps} />
                </Layout>}
              </MuiThemeProvider>
            </RuiThemeProvider>
          </ReactiveBase>
        </JssProvider>
      </ComponentsProvider>
    </Container>;
  }
}
