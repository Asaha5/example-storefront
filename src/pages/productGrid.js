import React, { Component, Fragment } from "react";
import PropTypes from "prop-types";
import { observer, inject } from "mobx-react";
import Helmet from "react-helmet";
import trackProductListViewed from "lib/tracking/trackProductListViewed";
import { productSearchInputProps } from "components/ProductSearchInput";
import ProductGrid, { searchResultCardProps } from "components/ProductGrid";
import get from "lodash.get";
import {
  DataSearch,
  ResultCard
} from "@appbaseio/reactivesearch";
import initReactivesearch from "@appbaseio/reactivesearch/lib/server";
import defaultCatalogQuery from "components/ProductGrid/defaultQuery";

/**
 * @name transformSearchRequest
 * @param {Object} request - the request
 * @return {Object} -  the transformed request
 */
function transformSearchRequest(request) {
  // console.log("====== transformSearchRequest(request)", JSON.stringify(request));

  let preference;
  let elasticQuery;
  if (Array.isArray(request)) {
    // On the server, the request is an Array
    [preference, elasticQuery] = request;
  } else {
    // "parse" ndjson - new line delimited JSON
    const parsedBody = request.body.split("\n");
    [preference, elasticQuery] = parsedBody;
  }

  elasticQuery = typeof elasticQuery === "string" ? JSON.parse(elasticQuery) : elasticQuery;

  // console.log("parsedESQuery", parsedESQuery);
  const { size, from: startFrom } = elasticQuery;

  const page = ((startFrom / size) || 0) + 1;

  // Get query from sensor component, i.e. SearchInput component
  let graphqlQuery = get(elasticQuery, "query.bool.must[0].bool.must.graphqlQuery");
  if (!graphqlQuery) {
    // Get query from a results component, i.e. ResultCard
    graphqlQuery = get(elasticQuery, "query.bool.must[0].bool.must[0].graphqlQuery");
    // Otherwise, use default initial query, "match_all: {}"
    if (!graphqlQuery) {
      graphqlQuery = defaultCatalogQuery;
    }
  }

  const transformedBody = {
    query: graphqlQuery.replace(/PAGE/, page).replace(/PER_PAGE/, size)
  };

  // console.log("********************transformedBody", transformedBody);
  // return transformedBody;
  if (typeof window === "undefined") {
    // Executing in Node
    return transformedBody;
  }

  request.body = JSON.stringify(transformedBody);
  return request;
}

export const reactivesearchSettings = {
  app: "reaction.cdc.reaction.catalog.json-gen1",
  transformRequest: transformSearchRequest,
  url: "http://graphql-proxy.api.reaction.localhost:9201"
};

@inject("routingStore", "uiStore")
@observer
class ProductGridPage extends Component {
  static propTypes = {
    catalogItems: PropTypes.array,
    catalogItemsPageInfo: PropTypes.object,
    initialGridSize: PropTypes.object,
    isLoadingCatalogItems: PropTypes.bool,
    routingStore: PropTypes.object,
    shop: PropTypes.shape({
      currency: PropTypes.shape({
        code: PropTypes.string.isRequired
      })
    }),
    tag: PropTypes.object,
    uiStore: PropTypes.shape({
      pageSize: PropTypes.number.isRequired,
      setPageSize: PropTypes.func.isRequired,
      setSortBy: PropTypes.func.isRequired,
      sortBy: PropTypes.string.isRequired
    })
  };

  static async getInitialProps({ req }) {
    // It is not perfect, but the only way we can guess at the screen width of the
    // requesting device is to parse the `user-agent` header it sends.
    const userAgent = req ? req.headers["user-agent"] : navigator.userAgent;
    const width = (userAgent && userAgent.indexOf("Mobi")) > -1 ? 320 : 1024;

    const searchStore = await initReactivesearch(
      [
        {
          ...productSearchInputProps,
          type: "DataSearch",
          source: DataSearch
        },
        {
          ...searchResultCardProps,
          type: "ResultCard",
          source: ResultCard
        }
      ],
      null,
      reactivesearchSettings,
    );

    return { initialGridSize: { width }, searchStore };
  }

  @trackProductListViewed()
  componentDidMount() {
    const { routingStore } = this.props;
    routingStore.setTag({});
  }

  componentDidUpdate(prevProps) {
    if (this.props.catalogItems !== prevProps.catalogItems) {
      this.trackEvent(this.props);
    }
  }

  @trackProductListViewed()
  trackEvent() {}

  setPageSize = (pageSize) => {
    this.props.routingStore.setSearch({ limit: pageSize });
    this.props.uiStore.setPageSize(pageSize);
  };

  setSortBy = (sortBy) => {
    this.props.routingStore.setSearch({ sortby: sortBy });
    this.props.uiStore.setSortBy(sortBy);
  };

  render() {
    const {
      shop
    } = this.props;
    const pageTitle = shop && shop.description ? `${shop.name} | ${shop.description}` : shop.name;

    return (
      <Fragment>
        <Helmet
          title={pageTitle}
          meta={[{ name: "description", content: shop && shop.description }]}
        />
        <ProductGrid />
      </Fragment>
    );
  }
}

export default ProductGridPage;
