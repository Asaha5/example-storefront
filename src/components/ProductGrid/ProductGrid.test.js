import React from "react";
import renderer from "react-test-renderer";
import { Provider } from "mobx-react";
import { MuiThemeProvider } from "@material-ui/core/styles";
import theme from "custom/reactionTheme";
import { ReactiveBase } from "@appbaseio/reactivesearch";
import { reactiveSearchSettings } from "../../pages/productGrid";
import ProductGrid from "./ProductGrid";
import initialSearchData from "./__mocks__/searchResults.mock";

const routingStore = {
  pathname: "tag",
  query: {
    slug: "test-tag",
    querystring: "?this&that"
  }
};

test("basic snapshot", () => {
  const component = renderer.create((
    <MuiThemeProvider theme={theme}>
      <Provider routingStore={routingStore}>
        <ReactiveBase
          {...reactiveSearchSettings}
          initialState={initialSearchData}
        >
          <ProductGrid />
        </ReactiveBase>
      </Provider>
    </MuiThemeProvider>
  ));
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
