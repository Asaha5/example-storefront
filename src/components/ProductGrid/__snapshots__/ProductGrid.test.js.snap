// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic snapshot 1`] = `
<div
  className="css-1py3auy e1n4b2jv0"
  style={Object {}}
>
  <div
    className="SkProductGrid-resultCard-1"
    style={Object {}}
  >
    <div
      className="css-ixn26d e1hwwpu60"
    >
      <p
        className="css-14pyrcn "
      >
        30
         results found in 
        1
        ms
      </p>
    </div>
    <div
      className="css-omf3o4 "
    >
      <a
        className="resultImage css-1p9dq2h e1cc0ogo1"
        href="/product/thankful-t-shirt"
        onClick={[Function]}
        rel={null}
        target="_self"
      >
        <div
          className="css-g6qnsb e1cc0ogo0"
          style={
            Object {
              "backgroundImage": "url(http://localhost:3000/undefined)",
            }
          }
        />
        <h2
          className="css-1pwe92q e1hnx4sx0"
        >
          <div>
            Thankful T-Shirt
            $24.99
          </div>
        </h2>
        <article
          dangerouslySetInnerHTML={
            Object {
              "__html": "Blessed Clothing",
            }
          }
        />
      </a>
    </div>
    <div
      className="css-11s5mbb "
    >
      <a
        className="css-dvxycr e165j7gc0"
        disabled={true}
        onClick={[Function]}
        onKeyPress={[Function]}
        tabIndex="0"
      >
        Prev
      </a>
      <a
        className=" active css-fdtpv0 e165j7gc0"
        href="?searchResultCard=1"
        onClick={[Function]}
        onKeyPress={[Function]}
        tabIndex="0"
      >
        1
      </a>
      <a
        className="css-4kq8mq e165j7gc0"
        href="?searchResultCard=2"
        onClick={[Function]}
        onKeyPress={[Function]}
        tabIndex="0"
      >
        2
      </a>
      <a
        className="css-4kq8mq e165j7gc0"
        href="?searchResultCard=3"
        onClick={[Function]}
        onKeyPress={[Function]}
        tabIndex="0"
      >
        3
      </a>
      <a
        className="css-4kq8mq e165j7gc0"
        href="?searchResultCard=4"
        onClick={[Function]}
        onKeyPress={[Function]}
        tabIndex="0"
      >
        4
      </a>
      <a
        className="css-4kq8mq e165j7gc0"
        href="?searchResultCard=5"
        onClick={[Function]}
        onKeyPress={[Function]}
        tabIndex="0"
      >
        5
      </a>
      <a
        className="css-4kq8mq e165j7gc0"
        disabled={false}
        href="?searchResultCard=2"
        onClick={[Function]}
        onKeyPress={[Function]}
        rel="next"
        tabIndex="0"
      >
        Next
      </a>
    </div>
  </div>
</div>
`;
